tensorflow_quantum/datasets/spin_system_test.py

```diff
@@ -0,0 +1,97 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Test the spin system dataset"""
+import tensorflow as tf
+import numpy as np
+import cirq
+from tensorflow_quantum.datasets import spin_system
+from tensorflow_quantum.datasets.spin_system import SpinSystemInfo
+
+
+class TFIChainTest(tf.test.TestCase):
+    """Testing tfi_chain."""
+
+    def test_errors(self):
+        """Test that it errors on invalid arguments."""
+        with self.assertRaisesRegex(ValueError,
+                                    expected_regex='Supported number of'):
+            qbs = cirq.GridQubit.rect(3, 1)
+            spin_system.tfi_chain(qbs)
+        with self.assertRaisesRegex(
+                ValueError, expected_regex='Supported boundary conditions'):
+            qbs = cirq.GridQubit.rect(4, 1)
+            spin_system.tfi_chain(qbs, 'open')
+        with self.assertRaisesRegex(TypeError,
+                                    expected_regex='expected str, bytes '):
+            qbs = cirq.GridQubit.rect(4, 1)
+            spin_system.tfi_chain(qbs, data_dir=123)
+        with self.assertRaisesRegex(TypeError,
+                                    expected_regex='must be a list of'):
+            spin_system.tfi_chain(['bob'])
+
+        with self.assertRaisesRegex(TypeError,
+                                    expected_regex='must be a one-dimensional'):
+            qbs = cirq.GridQubit.rect(4, 1)
+            spin_system.tfi_chain([qbs])
+
+    def test_fidelity(self):
+        """Test that all fidelities are close to 1."""
+        supported_nspins = [4, 8, 12, 16]
+        for nspins in supported_nspins:
+            qbs = cirq.GridQubit.rect(nspins, 1)
+            circuits, _, _, addinfo = spin_system.tfi_chain(
+                qbs,
+                'closed',
+            )
+            for n in range(len(addinfo)):
+                phi = cirq.Simulator().simulate(circuits[n]).final_state
+                gs = addinfo[n].gs
+                self.assertAllClose(np.abs(np.conj(np.dot(gs, phi))),
+                                    1.0,
+                                    rtol=1e-3)
+
+    def test_paulisum(self):
+        """Test that the PauliSum Hamiltonians give the ground state energy."""
+        supported_nspins = [4, 8, 12, 16]
+        for nspins in supported_nspins:
+            qbs = cirq.GridQubit.rect(nspins, 1)
+            circuits, _, pauli_sums, addinfo = spin_system.tfi_chain(
+                qbs, 'closed')
+            qubit_map = {qbs[i]: i for i in range(len(qbs))}
+            for n in range(len(pauli_sums)):
+                phi = cirq.Simulator().simulate(circuits[n]).final_state
+                e = pauli_sums[n].expectation_from_wavefunction(phi, qubit_map)
+                self.assertAllClose(e, addinfo[n].gs_energy, rtol=1e-4)
+
+    def test_returned_objects(self):
```

Comment:
Awesome! One final test that would be nice to see is that if I take a circuit from additional_info and resolve in the paramresolver do I get the original datapoint circuit ? This could be done via comparing the final wavefunctions for approximate equality.